### ISSUE TITLE
[RFC] Add a way to make Redis::subscribe() async

### ISSUE THREAD

**Issue by lyrixx** [phpredis/phpredis#1728]
### Expected behaviour

```php
<?php

$redis = new \Redis();
$redis->connect('172.17.0.2');

function f() {
    dump(func_get_args());
}

dump("START");

$redis->subscribe(['test'], 'f');

// Be able to run some code...
dump('we continue');
```

This is does not work since `Redis::subscribe()` is blocking.
It would be nice to have the same system as `pcntl_signal()`.

It would be even more awesome it this could work with signal:

<details>

```php
<?php

$redis = new \Redis();
$redis->connect('172.17.0.2');

function f() {
    dump(func_get_args());
}

dump("START");

pcntl_async_signals(true);
$shouldStop = false;
pcntl_signal(SIGINT, function () use (&$shouldStop) {
    dump('CTRL+C detected!');
    $shouldStop = true;

    exit(1);
});

$redis->subscribe(['test'], 'f');

while (true) {
    echo ".";
    if ($shouldStop) {
        echo "\n";
        return;
    }
    sleep(1);
}
```

</details>

### Actual behaviour

The following line blocks the script

```php
$redis->subscribe(['test'], 'f');
```

### I'm seeing this behaviour on
- OS: ubuntu
- Redis: 5+
- PHP: 7.4+
- phpredis: 5.2.0

### I've checked
- [x] There is no similar issue from other users
- [x] Issue isn't fixed in `develop` branch

**Comment by michael-grunder** [comment]
Adding support for non-blocking sockets would be non-trivial, although I can see the appeal.

We're going to be making some pretty substantial changes to the library to fully support Redis 6.0 and I'll keep non-blocking sockets in mind.

It may be possible to implement this kind of support _only_ for `subscribe` and `psubscribe`, which might make it more feasible to implement.

**Comment by merlinthemagic** [comment]
I am using this extension, but had to roll my own socket just for the pub/sub. This being a PHP extension i assume many are using it to maintain bi-directional sync to some frontend object model. Would be nice to have native support for servicing other requests and not hang blocked on subscriptions.

@michael-grunder Can you expand a bit on why this would be a non-trivial change? I would dig into i and get you a PR if its not a monumental task. I dont know this code base well so a few pointers on how wide you expect the blast radius extends and why. Not a tome just a few "obvious" areas of concern. I would want to limit the scope and only make the reading of subscription messages a non-blocking call rather than a block + call back.

**Comment by michael-grunder** [comment]
Hi @merlinthemagic,

I haven't taken a deep look at how specifically to tackle non-blocking functionality in PhpRedis, but it's non-trivial by nature.  There would need to be some way to either hook into PHP's event loop or into a third party one ([possibly using the libev extension](https://github.com/m4rw3r/php-libev)).

I do think it's worth looking at for PhpRedis 6 since we're already going to be replacing all the RESP parsing machinery with hiredis.

**Comment by lyrixx** [comment]
Maybe, we should have a look at pcntl, and see how this extension is implemented. Because it support some "interuption"

**Comment by nikeee** [comment]
Is there any news regarding this issue? I'm currently replacing RabbitMQ with Redis and I'm facing the issue that subscribing blocks.

On RabbitMQ, you can call `$channel->wait()` to block the thread. However, this also calls all callbacks of different subscriptions if there are any. So it can be used to process different channels instead of just one like in phpredis. Basically something like this:
```php
$redis = new Redis();

echo 'Starting';

$redis->subscribe(['a'], function($redis) {
	$redis->lpop('c');
});
$redis->subscribe(['b'], 'g');

echo 'Waiting for input...';
$redis->wait(); // blocks, but will call f and g if there are any messages
```

At the moment, it's also not possible to do any other redis operations in the callbacks. However, it is a common pattern to listen for keypsace changes in a pub/sub and issue other redis commands reacting to that.

I think this model is far easier to implement than using interrupts (though, it is less powerful). It also does not require an additional event-loop runtime.

**Comment by michael-grunder** [comment]
Unfortunately, you can't really do what you're looking for unless and until we implemented logic for it.  One problem is that PhpRedis uses RESP3 which prohibits executing other commands while subscribed.

This limitation was removed with RESP3's `push` reply type, but we don't currently support that protocol version.

**Comment by nikeee** [comment]
What is needed to support the push type? Whats the statos of the hiredis migration (mentioned above)?

**Comment by nikeee** [comment]
I just tried [predis](https://github.com/predis/predis) and it seems to have the same issue. However, I found a solution to work with redis in a callback: Just create a second redis client and use that for that operation.
This was the solution for predis, this might also work for phpredis. If so, consider my request as superseded.

**Comment by sergiuionescu** [comment]
Maybe a resource approach would fit the bill better api here?

```
$subscriptionResource = $redis->subscribe(['a']);

$message = $subscriptionResource->fetch();// blocking

yield $message;
```

I am trying to stream messages from a backend socket connection to mutiple clients, and being able to use a generator here would be significantly better than being stuck in the subscribe callback loop.

**Comment by Nurlan270** [comment]
> I just tried [predis](https://github.com/predis/predis) and it seems to have the same issue. However, I found a solution to work with redis in a callback: Just create a second redis client and use that for that operation. This was the solution for predis, this might also work for phpredis. If so, consider my request as superseded.

Can you give an example how to do it in predis?

**Comment by nikeee** [comment]
@Nurlan270 it's something like
```php
$redis = new Redis();
$redis->connect(...);

$subscriberRedis = new Redis();
$subscriberRedis->connect(...);

$subscriberRedis->subscribe('a', function($message) {
	echo "Received message: $message\n";

	$redis->lPush('b', $message);
});
```

`$subscriberRedis->subscribe` blocks (and $loopClient cannot be used inside the callback), but you can still use `$redis` inside.
